dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/SolrSearchResultsService.java

```diff
@@ -0,0 +1,327 @@
+            if (result.getDatasetVersionId() == null
+                    || (datasetData = datasets.get(result.getDatasetVersionId().intValue())) == null) {
+                continue;
+            }
+            if ("DEACCESSIONED".equals(datasetData[DATASETS_QUERY_VERSION_STATE])) {
+                result.setDeaccessionedState(true);
+            }
+            setIfNotNull((String) datasetData[DATASETS_QUERY_DATAVERSE_ALIAS], result::setDataverseAlias);
+            setIfNotNull(result.getDeaccessionReason(), result::setDescriptionNoSnippet);
+
+            Dataset entity = createAndFillDataset(result.getIdentifier(), datasetData, thumbnails);
+            result.setEntity(entity);
+        }
+    }
+
+    public void populateDatafileSearchCard(Collection<SolrSearchResult> solrResults) {
+        Set<Long> ids = extractNonNullValuesSet(solrResults, SolrSearchResult::getEntityId);
+        Map<Long, DataFile> datafiles = findDatafiles(ids);
+        for (SolrSearchResult result : solrResults) {
+            DataFile datafile;
+            if ((datafile = datafiles.get(result.getEntityId())) != null) {
+                result.setEntity(datafile);
+            }
+        }
+    }
+
+    public Map<Long, DataFile> findDatafiles(Collection<Long> ids) {
+        Collection<Object[]> rawFilesData = callNamedNativeQueryWithIds(DATAFILES_QUERY_BASE_NAME, ids);
+        Set<Integer> idsWithTags = rawFilesData.stream()
+                .filter(d -> TextMimeType.TSV.getMimeValue().equalsIgnoreCase((String) d[DATAFILES_QUERY_FILE_CONTENTTYPE]))
+                .map(d -> (Integer) d[DATAFILES_QUERY_FILE_ID])
+                .collect(Collectors.toSet());
+        Map<Long, List<Integer>> datafileTags = callNamedNativeQueryWithIds(DATAFILETAGS_QUERY_BASE_NAME, idsWithTags).stream()
+                .collect(Collectors.groupingBy(d -> (Long) d[DATAFILETAGS_QUERY_DATAFILE_ID],
+                        Collectors.mapping(d -> (Integer) d[DATAFILETAGS_QUERY_TYPE], Collectors.toList())));
+        List<String> tagLabels = DataFileTag.listTags();
+
+        Map<Long, DataFile> datafiles = new HashMap<>();
+        for (Object[] fileData : rawFilesData) {
+            DataFile dataFile = createAndFillDataFile(fileData);
+            datafiles.put(dataFile.getId(), dataFile);
+            fillTabularDataIfNeeded(dataFile, fileData, datafileTags, tagLabels);
+        }
+        return datafiles;
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private <T, U> Set<U> extractNonNullValuesSet(Collection<T> source, Function<T, U> valueMapper) {
+        return source != null
+                ? source.stream()
+                    .map(valueMapper)
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toSet())
+                : Collections.emptySet();
+    }
+
+    private Collection<Object[]> callNamedNativeQueryWithIds(String queryBaseName, Collection<? extends Number> ids) {
```

Comment:
Can we add a comment which describes what we are doing here and some link that explains why?